exe/Rules.hs

```diff
@@ -0,0 +1,150 @@
+module Rules
+  ( loadGhcSession
+  , cradleToSession
+  , cradleLoadedMethod
+  , createSession
+  , getComponentOptions
+  )
+where
+
+import           Control.Exception
+import           Control.Monad                  (filterM)
+import qualified Crypto.Hash.SHA1               as H
+import           Data.ByteString.Base16         (encode)
+import qualified Data.ByteString.Char8          as B
+import           Data.Functor                   ((<&>))
+import           Data.Maybe                     (fromMaybe)
+import           Data.Text                      (Text)
+import           Development.IDE.Core.Rules     (defineNoFile)
+import           Development.IDE.Core.Shake     (sendEvent, define, useNoFile_)
+import           Development.IDE.GHC.Util
+import           Development.IDE.Types.Location (fromNormalizedFilePath)
+import           Development.Shake
+import           DynFlags                       (gopt_set, gopt_unset,
+                                                 updOptLevel)
+import           GHC
+import qualified GHC.Paths
+import           HIE.Bios
+import           HIE.Bios.Cradle
+import           HIE.Bios.Environment           (addCmdOpts)
+import           HIE.Bios.Types
+import           Linker                         (initDynLinker)
+import           RuleTypes
+import qualified System.Directory.Extra         as IO
+import           System.Environment             (lookupEnv)
+import           System.FilePath.Posix          (addTrailingPathSeparator,
+                                                 (</>))
+import Language.Haskell.LSP.Messages as LSP
+import Language.Haskell.LSP.Types as LSP
+import Data.Aeson (ToJSON(toJSON))
+
+-- Prefix for the cache path
+cacheDir :: String
+cacheDir = "ghcide"
+
+notifyCradleLoaded :: FilePath -> LSP.FromServerMessage
+notifyCradleLoaded fp =
+    LSP.NotCustomServer $
+    LSP.NotificationMessage "2.0" (LSP.CustomServerMethod cradleLoadedMethod) $
+    toJSON fp
+
+loadGhcSession :: Rules ()
+loadGhcSession =
+    -- This rule is for caching the GHC session. E.g., even when the cabal file
+    -- changed, if the resulting flags did not change, we would continue to use
+    -- the existing session.
+    defineNoFile $ \(GetHscEnv opts deps) ->
+        liftIO $ createSession $ ComponentOptions opts deps
+
+cradleToSession :: Rules ()
+cradleToSession = define $ \LoadCradle nfp -> do
+    let f = fromNormalizedFilePath nfp
+
+    -- If the path points to a directory, load the implicit cradle
+    mbYaml <- doesDirectoryExist f <&> \isDir -> if isDir then Nothing else Just f
+    cradle <- liftIO $  maybe (loadImplicitCradle $ addTrailingPathSeparator f) loadCradle mbYaml
+
+    sendEvent $ notifyCradleLoaded f
```

Comment:
I would prefer to hide this behind a flag that we pass only in the tests. Especially if the message does not start with `$/`, clients might fall over if they see this message and don’t understand it and even if we prefix it with `$/` it’s really something that is only intended for tests.

Reply:
Done!